AUTO_MAX_PARTITION

JackZhouVCD opened the pull request:
## Description
This PR Implements a shortcut mechanism to determine if search depths could be skipped based on motion estimation info and other statistics.

The mechanism uses a feed-forward Neural Network with ReLU hidden-layer activations and identity output activation. The inference implementation and model weights are both ported from aomenc

This feature is on by default, and it is not supported for high bit-depth modes (anything beyond 8 bit bit-depth). A new API switch `-enable-amp` is introduced to turn this feature either on or off.

closes #799 

## Author

@hguermaz 
@JackZhouVCD 

## Type of change

Enhancement

## Tests and performance
All data generated from full-list objective-1-fast dataset in the context of enc-mode 0. Speed data is generated using AWS ec2 48 vCPU instances

| | Deviation|
|-|-|
| **BD-Rate** | 0.026% |
| **Speed** | 3.507% |

hassount commented:
@joelsole @mafonsoN any issues merging this ?

hassount commented:
@joelsole @mafonsoN any issues merging this ?

joelsole commented:
@andrey-norkin is reviewing it.

hassount commented on Source/Lib/Common/Codec/aom_dsp_rtcd.h:
```diff
@@ -2908,6 +2910,12 @@ extern "C" {
 
+#if AUTO_MAX_PARTITION
+    void av1_nn_predict_c(const float *input_nodes, const NN_CONFIG *const nn_config, int reduce_prec, float *const output);
+    void av1_nn_predict_sse3(const float *input_nodes, const NN_CONFIG *const nn_config, int reduce_prec, float *const output);
```
@ttrigui can you please check if this would benefit from an AVX2 version?

hassount commented:
This is up next in the P1 list. 
@JacksonDagger can you please rebase to the latest master and run some verification tests to make sure everything checks-out. 
@andrey-norkin Can we assume that this PR is ok, pending the comments being resolved ?

joelsole commented:
@hassount  this PR introduces ~6500 lines of code. 
The benefit seems to be a 3% encoding time decrease for mode 0. Is there any other benefit? 
How does this interact with the recomputation of the encoding modes?

hassount commented:
@joelsole I ll let @hguermaz respond to this

JackZhouVCD commented on Source/API/EbSvtAv1Enc.h:
```diff
@@ -462,6 +462,8 @@ typedef struct EbSvtAv1EncConfiguration
 
     uint32_t                     sq_weight;
+
+    uint8_t                     enable_auto_max_partition;
```
This will be addressed when I rebase my changes

hassount commented:
@JackZhouVCD 
Can  you also please create an issue that this PR is addressing?

JackZhouVCD commented on Source/Lib/Common/Codec/EbProductCodingLoop.c:
```diff
@@ -8285,6 +8294,285 @@ void update_skip_next_nsq_for_a_b_shapes(
+#define FLT_MAX          3.402823466e+38F        // max value
+#define FEATURE_SIZE_MAX_MIN_PART_PRED 13
+
+uint32_t mds_idx_16x16[64] = {
```
this will be addressed when I rebase my changes

hguermaz commented:
@joelsole we integrated all libaom neural networks/machine learning files into SVT (i.e. not just what we need for auto max partition) as we are planning to adopt more shortcuts that use neural networks. @JackZhouVCD please go ahead and remove useless materials until adoption of more shortcuts.

JackZhouVCD commented on Source/Lib/Common/Codec/EbProductCodingLoop.c:
```diff
@@ -8285,6 +8294,285 @@ void update_skip_next_nsq_for_a_b_shapes(
+    uint16_t             sb_origin_x,
+    uint16_t             sb_origin_y) {
+    int f_idx = 0;
+    uint32_t qIndex = picture_control_set_ptr->parent_pcs_ptr->frm_hdr.quantization_params.base_q_idx;
```
this will be addressed when I rebase my changes

JackZhouVCD commented on Source/Lib/Common/Codec/EbEncDecProcess.c:
```diff
@@ -1560,6 +1560,13 @@ EbErrorType signal_derivation_enc_dec_kernel_oq(
 
+#if AUTO_MAX_PARTITION
+    // signal for enabling shortcut to skip search depths
+    if (picture_control_set_ptr->parent_pcs_ptr->sequence_control_set_ptr->static_config.encoder_bit_depth > EB_8BIT)
```
The inputs to the neural network that this feature uses contain components that are assumed to be 8-bit. For example variance information and motion compensation for 16x16 blocks.

JackZhouVCD commented:
> @JackZhouVCD can you please rebase to the latest master and run some verification tests to make sure everything checks-out.

@hassount rebased and addressed the comments.

> @hassount this PR introduces ~6500 lines of code.
> The benefit seems to be a 3% encoding time decrease for mode 0. Is there any other benefit?
> How does this interact with the recomputation of the encoding modes?

@joelsole @hguermaz roughly 5000 of those lines were from weights and biases in neural networks  that we were not using and a V2 infrastructure that libaom has yet to adopt themselves. I have removed those now so the PR is a lot lighter.


> @JackZhouVCD
> Can you also please create an issue that this PR is addressing?

@hassount sure I will open an issue shortly

JackZhouVCD commented:
rebased again since i saw there were merge conflicts

andrey-norkin commented on Source/Lib/Common/Codec/aom_dsp_rtcd.h:
```diff
@@ -2919,6 +2921,12 @@ extern "C" {
     RTCD_EXTERN void(*compute_interm_var_four8x8)(uint8_t *input_samples, uint16_t input_stride, uint64_t *mean_of8x8_blocks, uint64_t *mean_of_squared8x8_blocks);
 
+#if AUTO_MAX_PARTITION
+    void av1_nn_predict_c(const float *input_nodes, const NN_CONFIG *const nn_config, int reduce_prec, float *const output);
```
Is there any unit test that compares output of these two functions, c and sse versions?

JackZhouVCD replied:
There is a unit test that does that now

JackZhouVCD commented on Source/App/EncApp/EbAppConfig.c:
```diff
@@ -340,6 +341,8 @@ static void SetMDS2_PRUNE_S_TH(const char *value, EbConfig *cfg) {
         cfg->md_stage_2_cand_prune_th = (uint64_t)~0;
 }
 
+static void SetEnableAutomaxPartition           (const char *value, EbConfig *cfg) { cfg->enable_auto_max_partition = (uint8_t)strtol(value, NULL, 0); };
```
good point, I have fixed that now. As a side note I think we should take time to update all of the function names in this section to be snake cased. As of the moment we have a mix between the two in this section

hassount replied:
agreed #668

andrey-norkin commented:
Approved. 
Need to fix the failing builds though

JackZhouVCD commented:
> Approved.
> Need to fix the failing builds though

Thanks. I'm not sure if the CI failing is to do with this PR though, the [style check is the one that's not passing](https://travis-ci.com/OpenVisualCloud/SVT-AV1/jobs/264205385#L1232) and it's pointing to malformed commit messages in the past (at the latest june)
